### PR TITLE
NO-ISSUE: Push cluster_installation_failed event each time the state …

### PR DESCRIPTION
**Pull request by pastequo**
…machine go to failed state

Leveraging PostTransition functions.

The idea here is to start isolating the events logic aside from the state logic

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by pastequo** [comment]
cc @rccrdpccl

**Review by eranco74 — COMMENTED** [review]
/lgtm

File: internal/cluster/transition.go [hunk 1]
```diff
@@ -73,6 +73,45 @@ type TransitionHandler interface {
...
+
+type EventHandler interface {
+	SendErrorEvent(template string, formatFuncs ...formatFunction) stateswitch.PostTransition
+}
+
+type ErrorEventArgs interface {
+	GetContext() context.Context
+}
+
+func (th *transitionHandler) SendErrorEvent(template string, formatFuncs ...formatFunction) stateswitch.PostTransition {
```
Comment (paul-maidment) [inline internal/cluster/transition.go:92]: I would argue that this function should be called `SendClusterInstallationErrorEvent` or even `SendClusterInstallationFailedEvent` as 

* this is not some abstract event sender that may apply to any kind of event, this is specific to Cluster Installation
* ClusterInstallation Failure is not the only type of failure event handled in this file.

so `SendErrorEvent` does not unambiguously represent what this function does.

File: internal/cluster/statemachine.go [hunk 2]
```diff
@@ -14,6 +15,21 @@ const (
 	TransitionTypeRefreshStatus          = "RefreshStatus"
 )
 
+func SequentialPostTransitions(transitions ...stateswitch.PostTransition) stateswitch.PostTransition {
+	return func(stateSwitch stateswitch.StateSwitch, args stateswitch.TransitionArgs) error {
+		var ret error
+
+		for _, transition := range transitions {
```
Comment (paul-maidment) [inline internal/cluster/statemachine.go:22]: I think we should use the term `postTransition` and `postTransitions` here
as we should make it clear to the reader that these are not actually transitions.

**Pull request merged**